Add typing information for contexts, token

mrnicegyu11 opened the pull request:
# Description

Trying to get https://github.com/open-telemetry/opentelemetry-python/pull/4164 in.

Opening this 1:1 similar PR for now, with changes from upstream/main, to see how the tests look. Locally, all tests are green.

~~This is WIP for now.~~

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Ran the `tox` full test suite locally, all green. Waiting for maintainer approval for the github actions tests

# Does This PR Require a Contrib Repo Change?

<!--
Answer the following question based on these examples of changes that would require a Contrib Repo Change:
- [The OTel specification](https://github.com/open-telemetry/opentelemetry-specification) has changed which prompted this PR to update the method interfaces of `opentelemetry-api/` or `opentelemetry-sdk/`
- The method interfaces of `test/util` have changed
- Scripts in `scripts/` that were copied over to the Contrib repo have changed
- Configuration files that were copied over to the Contrib repo have changed (when consistency between repositories is applicable) such as in
    - `pyproject.toml`
    - `isort.cfg`
    - `.flake8`
- When a new `.github/CODEOWNER` is added
- Major changes to project information, such as in:
    - `README.md`
    - `CONTRIBUTING.md`
-->

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

- [x] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

Kludex commented on opentelemetry-api/tests/context/base_context.py:
```diff
@@ -64,7 +64,7 @@ def test_attach(self):
 
             with self.assertLogs(level=ERROR):
-                context.detach("some garbage")
+                context.detach("some garbage")  # type:ignore
```
Why is the `type: ignore` needed?

```suggestion
                context.detach("some garbage")  # type: ignore[?]
```

mrnicegyu11 replied:
True thanks this was actually not necessary, slipped past me. I removed the `type:ignore`

Kludex reviewed (changes requested):
The `opentelemetry/context` directory needs to be included in the mypy setup.

---

Not related to this PR, but I already pointed out that this setup doesn't make sense. mypy needs to run once, and it should have the context of the local imports.

I've shared my thoughts here: https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3116#issuecomment-2557148675

I'm happy to push a PR switching the type checker used here, and I'm happy to help this project to have proper type hints. Otherwise, I think someone should try to make the mypy setup work properly.

mrnicegyu11 commented:
> The `opentelemetry/context` directory needs to be included in the mypy setup.
> 
> Not related to this PR, but I already pointed out that this setup doesn't make sense. mypy needs to run once, and it should have the context of the local imports.
> 
> I've shared my thoughts here: [open-telemetry/opentelemetry-python-contrib#3116 (comment)](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3116#issuecomment-2557148675)
> 
> I'm happy to push a PR switching the type checker used here, and I'm happy to help this project to have proper type hints. Otherwise, I think someone should try to make the mypy setup work properly.
> 
> I'll bring the above to the next Python SIG meeting.

Thanks for the review, as some other people before me I am implementing opentelemetry in code I am working on, and I have type-checking for this code setup which accounts for dependencies. There was a bug (on my side, wrongly using open-telemetry) that could have been avoided or would have been caught by `opentelemetry-python` providing the correct types. For this reason I have decided to PR this small atomic change (again). This is just to spare some pain for successors, at this point I was not planning to involve myself in overhauling the whole typechecking in this opentelemtry-python-repo :sweat_smile: @Kludex 

From where I am sitting, merging this small PR would bring immediate improvements for everyone that runs a type-checker on their code and imports from opentelemetry-python. I am not sure what the procedure is, who to talk to and such, to get it in. @hyoinandout actually already opened a PR with these exact changes ( https://github.com/open-telemetry/opentelemetry-python/pull/4164 ), but I guess they could not get it merged. 

Any advice or support would be welcome from my side, as I am fairly sure other people will run into issues that could be prevented by proper typing here :--)

mrnicegyu11 commented:
Or, well, if any maintainer feels that realistically this PR wont be merged, then please let me know as well, to spare some work&time ;)

Kludex commented:
The `opentelemetry/context` directory needs to be included in the mypy setup.

mrnicegyu11 commented:
Hey @Kludex and others , I hope I am not stealing too much of your time thanks for the consideration :--)

I read up on your pyright vs. mypy issue mentioned, and checked my PR here for this and your comment to include the `opentelemetry/context` folder in mypy.

From what I can see:
- The `mypy` that can be run in *this repo* via `tox -e mypy` does include the files I have changed in this PR, and specifically checks the files in `opentelemetry-api/src/opentelemetry/context/*`. So the types of these modified files are being correctly checked, as per the current setup at least.
- There was another typing bug in my PR (:innocent:), which is now detected and fixed (the previous `context.detach(string)` has wrong types, one can only detach tokens)
- I saw that your recently changed the `mypy.ini` so I am not sure if you might have adressed your previous concerns before? 

Summarizing, w.r.t. `The opentelemetry/context directory needs to be included in the mypy setup.`, it seems that this is the case in this repo (unless I am still not fully following, maybe you are referring to other related repositories of opentelemetry?). The `tox` suite tests run through 100% green locally for me, for this PR. Would you reconsider this typing contribution with this info?


Thanks a lot, Cheers!

Kludex reviewed (approved):
LGTM!

Kludex commented:
Sorry for being annoying. This looks great. Thanks! :)

lzchen reviewed (approved):
@mrnicegyu11 

Thanks a bunch for the contribution and apologies for the late review!

emdneto commented on opentelemetry-api/src/opentelemetry/context/__init__.py:
```diff
@@ -14,6 +14,7 @@
 
```
```suggestion

from __future__ import annotations

```

emdneto commented on opentelemetry-api/src/opentelemetry/context/context.py:
```diff
@@ -14,6 +14,7 @@
 
```
```suggestion

from __future__ import annotations

```

emdneto commented on opentelemetry-api/src/opentelemetry/context/contextvars_context.py:
```diff
@@ -11,7 +11,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from contextvars import ContextVar
+from contextvars import ContextVar, Token
```
```suggestion

from __future__ import annotations

from contextvars import ContextVar, Token
```

emdneto commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Unreleased
 
+- Add type annotations to context's attach & detach
+  ([#4164](https://github.com/open-telemetry/opentelemetry-python/pull/4346))
```
```suggestion
  ([#4346](https://github.com/open-telemetry/opentelemetry-python/pull/4346))
```

emdneto reviewed (commented):
Left some changes that I think will make tests pass, otherwise won't work for py38

xrmx commented on docs/conf.py:
```diff
@@ -150,6 +150,10 @@
     ),
+    (
+        "py:class",
+        "_contextvars:Token",
```
```suggestion
        "contextvars:Token",
```

xrmx commented on docs/conf.py:
```diff
@@ -150,6 +150,10 @@
         "py:class",
         "opentelemetry.sdk.metrics._internal.aggregation._Aggregation",
     ),
+    (
```
This is not working :disappointed:

xrmx commented on docs/conf.py:
```diff
@@ -150,6 +150,10 @@
     ),
+    (
+        "py:class",
+        "contextvars:Token",
```
```suggestion
        "_contextvars.Token",
```
